execute/report/report.go

```diff
@@ -149,14 +129,62 @@ const (
 )
 
 func (b *execReportBuilder) checkMessage(
-	_ context.Context, idx int, execReport plugintypes.ExecutePluginCommitData,
+	ctx context.Context, idx int, execReport plugintypes.ExecutePluginCommitData,
 	// TODO: get rid of the nolint when the error is used
-) (messageStatus, error) { // nolint this will use the error eventually
-	if slices.Contains(execReport.ExecutedMessages, execReport.Messages[idx].Header.SequenceNumber) {
-		return AlreadyExecuted, nil
+) (plugintypes.ExecutePluginCommitData, messageStatus, error) { // nolint this will use the error eventually
+	if idx >= len(execReport.Messages) {
+		b.lggr.Errorw("message index out of range", "index", idx, "numMessages", len(execReport.Messages))
+		return execReport, TokenDataFetchError, fmt.Errorf("message index out of range")
+	}
+
+	msg := execReport.Messages[idx]
+
+	// Check if the message has already been executed.
+	if slices.Contains(execReport.ExecutedMessages, msg.Header.SequenceNumber) {
+		b.lggr.Infow(
+			"message already executed",
+			"messageID", msg.Header.MessageID,
+			"sourceChain", execReport.SourceChain,
+			"seqNum", msg.Header.SequenceNumber)
+		return execReport, AlreadyExecuted, nil
+	}
+
+	// Check if token data iss ready.
```

Comment:
```suggestion
	// Check if token data is ready.
```

We're not in the ISS ;-)